joomla/joomla-cms
# Add versioning to the modules

### Summary of Changes
This plugin adds the option to use module versions. Every time a module is saved, a new version is stored in the database. 

![Image-001](https://user-images.githubusercontent.com/5610413/199287435-77a6140d-bc28-40c2-9704-143cc3e45c8f.jpg)

Stored versions can be restored by selection the version you want to restore and clicking the Restore button. The current version is marked with a star. If a version is restored and again saved, then this will be a new version.

![Image-002](https://user-images.githubusercontent.com/5610413/199288527-1e24b27e-e689-4e03-9031-33035ab7df88.jpg)

The module parameters can be viewed in an accordion as raw output for comparison. 

![Image-003](https://user-images.githubusercontent.com/5610413/199288886-d09d1b4c-92d9-40ae-8a88-cfd0528a87f4.jpg)

In the plugin parameters there are two options. One to disable the parameter view and one to set the number of versions.

![Image-004](https://user-images.githubusercontent.com/5610413/199289189-46e6a80e-76bf-4c59-b5d5-d11ed12cf15f.jpg)

When a module is deleted, also the corresponding versions for that module will be deleted.

Note: The is some styling added, so a **npm ci** for building the stylesheets is required,

### Testing Instructions
1. Install a module for testing
2. Save the module for the first time. The version button is now visible,
3. Make some changes to the module and store again.
4. Compare the versions via the tabs
5. Restore a stored version and check if all settings are restored. (User Access is not stored in the module so can't be restored).
6. Delete the module a check the database of all corresponding versions of that module are deleted.

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: Will be created after this PR is merged.

- [x] Pull Request link for manual.joomla.org: Will be created after this PR is merged.
